In [1]:
from google.colab import drive
drive.mount('/content/drive/')
#/content/drive/MyDrive/batches

Mounted at /content/drive/


In [2]:
#from libs.corpus import openConllu, check_projectivity
!pip install pyconll
import pyconll
import pyconll.util
import torch
from transformers import AutoModel, AutoTokenizer
import numpy as np
from torch.nn.utils.rnn import pack_sequence, unpack_sequence
import os

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print( torch.cuda.is_available())
print(device)

True
cuda


In [3]:

train = pyconll.load_from_file('/content/drive/MyDrive/magistrale/TLN/data/it_isdt-ud-train.conllu')
train_prepocesed=[]
for i,sent in enumerate(train):
    sentence_preprocesed=[]
    for j,token in enumerate(sent):
        if(token.head is not None):
            sentence_preprocesed.append(token)
    train_prepocesed.append(sentence_preprocesed)



In [87]:





#0:SHIFT 1:RIGHTARC 2:LEFTARC
moves = [0, 1, 2]


#creiamo un oggetto Dependencies per salvare le dependencies
class Dependencies(object):
    def __init__(self,n):
        self.n = n
        self.heads = [None] * (n+1)
        self.arcs = []

    def get_heads(self):
        return self.heads

    def add_arc(self, head, child):
        child=child
        self.heads[child]=head
        self.arcs.append((head,child))

    def contains(self,head,child):
        child=child
        if self.heads[child]==head:
            return True
        else: return False





class Oracle(object):

    def __init__(self) -> None:
        #BERT encoder
        encoder_name = "dbmdz/bert-base-italian-xxl-cased"
        self.tokenizer = AutoTokenizer.from_pretrained(encoder_name)
        self.encoder = AutoModel.from_pretrained(encoder_name)

        #LSTM oracle
        input_size = 4608
        hidden_size = 256
        num_layers = 1
        output_size = 3
        self.epoch = 12
        self.model = torch.nn.LSTM(input_size,hidden_size,num_layers,batch_first=True,bidirectional=False,proj_size=output_size).to(device)

        #Oracle critenion and optimazation
        self.criterion = torch.nn.CrossEntropyLoss()
        self.optimizer = torch.optim.Adam(self.model.parameters(), lr=0.0005)

    def tokens(self,words,lemmas):
        words_token=[102]
        for i in range(0,len(lemmas)):
            word_token=self.tokenizer.convert_tokens_to_ids(words[i])
            lemma_token=self.tokenizer.convert_tokens_to_ids(lemmas[i])
            if(word_token!=101):
                words_token.append(word_token)
            else:
                words_token.append(lemma_token)
        words_token.append(103)
        return words_token

    def encode(self,words,lemmas):
        words_token=self.tokens(words,lemmas)
        #max_length = 64  # Example max length ????
        padded_input_ids = words_token
        input_tensor = torch.tensor([padded_input_ids])
        with torch.no_grad():
            outputs = self.encoder(input_tensor)
        return outputs.last_hidden_state

    #prende i batches dalla cartella, e allena il modello
    def train_on_batches(self):
        files = os.listdir("/content/drive/MyDrive/magistrale/TLN/batches")
        for k in range(0,self.epoch):
          for i in range(0,len(files)):
              if(i<30):

                  print("Loadind "+str(i))
                  batch_features,batch_moves=torch.load(f"/content/drive/MyDrive/magistrale/TLN/batches/tensor{i}.pt")
                  batch_features=unpack_sequence(batch_features)
                  batch_moves=unpack_sequence(batch_moves)
                  for j in range(0,len(batch_moves)):
                      predicted_output, _ = self.model(batch_features[j].to(device))
                      loss = self.criterion(predicted_output, batch_moves[j].to(device))
                      self.optimizer.zero_grad()
                      loss.backward()
                      self.optimizer.step()


    def score(self,features):
        predicted_output,_ = self.model(features.to(torch.float32).to(device))
        return predicted_output

    #prende gli ultimi tre elementi dello stack
    def get_stack_context(self,list):
        depth=len(list)

        if depth >= 3:
            return [list[-1], list[-2], list[-3]]

        elif depth >= 2:

            return [list[-1], list[-2], -1]

        elif depth == 1:
            return [list[-1], -1 , -1]
        else:
            return [-1, -1, -1]

    #prende gli ultimi due elementi dell buffer
    def get_buffer_context(self,index,len_phrase):
        if(index==len_phrase-1):
            return [index,index+1,-1]
        elif(index==len_phrase):
            return [index,-1,-1]
        elif(index>len_phrase):
            return [-1,-1,-1]
        else: return [index,index+1,index+2]

    def flatten_embedded_features(self,matrix):
        flat_list = torch.tensor([])
        for row in matrix:
            flat_list=torch.cat((flat_list,row))
        return flat_list

    def extract_features(self,phrase,phrases_lemma,stacks,buffers):
        embeddings = self.encode(phrase,phrases_lemma)
        root_embeding=embeddings[0][0]
        empty_embedding=torch.tensor(np.zeros(768))
        embedded_features=[]
        for i in range(0,len(stacks)):
            stack_feature=self.get_stack_context(stacks[i])
            for j,el in enumerate(stack_feature):
                if(el>=1):
                    stack_feature[j]=embeddings[0][el]
                elif(el==0):
                    stack_feature[j]=root_embeding
                else:
                    stack_feature[j]=empty_embedding
            buffer_feature=self.get_buffer_context(buffers[i],len(phrase))
            for j,el in enumerate(buffer_feature):
                if(el>=1):
                    buffer_feature[j]=embeddings[0][el]
                else:
                    buffer_feature[j]=empty_embedding
            embedded_features.append(self.flatten_embedded_features(stack_feature+buffer_feature))
        return embedded_features



class Parser(object):
    def __init__(self,oracle):
        self.oracle=oracle

    #applica la mossa andando ad aggiornare lo stack e l'indice del buffer
    def transition(self,move, stack, i, dependencies):
        match move:
            case 0:
                stack.append(i)
                return stack,i+1,dependencies
            case 1:
                dependencies.add_arc(stack[-2], stack.pop())
                return stack,i,dependencies
            case 2:
                dependencies.add_arc(stack[-1], stack[-2])
                stack.pop(-2)
                return stack,i,dependencies
            case _:
                raise "Wrong Move"

    #ritorna le mosse possibili che si possono applicare
    def get_valid_moves(self,i, n, stack_depth):
        moves = []
        if i <= n:
            moves.append(0)
        if stack_depth >= 2:
            moves.append(1)
            moves.append(2)
        return moves

    def parsing(self,words,phrases_lemma):
        n=len(words)
        deps=Dependencies(n)
        stack=[0]
        i_buffer=1
        moves=self.get_valid_moves(i_buffer,n,len(stack))
        old_stack=[]
        old_buffer=[]
        memory=1
        while moves:
            features = self.oracle.extract_features(words,phrases_lemma,[stack],[i_buffer])
            features=torch.stack(features)
            scores = self.oracle.score(features)
            scores=scores[-1].tolist()
            next_move = max(moves, key=lambda move: scores[move])
            stack,i_buffer,deps = self.transition(next_move, stack, i_buffer, deps)
            moves = self.get_valid_moves(i_buffer,n,len(stack))

            if(len(old_stack)<memory):
                old_stack.append(stack)
                old_buffer.append(i_buffer)
            else:
                old_stack.pop(0)
                old_buffer.pop(0)
                old_stack.append(stack)
                old_buffer.append(i_buffer)

        return deps

    #sceglie la mossa migliore da eseguire nel simulate_parse
    def check_best(self,heads,stack,buffer,deps,i):
        move=-1
        if(len(stack)>=2):
            children_list=[]
            for child,head in enumerate(heads):
                if head == stack[-1]:
                    children_list.append(child)
            if(heads[stack[-2]]==stack[-1]):
                move=2
            if(((heads[stack[-1]])==(stack[-2])) and all([deps.contains(stack[-1],child) for child in children_list])):
                move=1
        if(i<=len(buffer) and move==-1):
            move=0
        elif(i>len(buffer) and move==-1):
            move=None
        return move
    #fa reverse engineering, dato lo stato finale ricostruisce lo stack, buffer e le mosse
    def simulate_parse(self,heads,buffer):
        deps=Dependencies(len(buffer))
        stack=[0]
        moves=[]
        buffers=[]
        stacks=[]
        i=1
        best_move=self.check_best(heads,stack,buffer,deps,i)
        while best_move!=None:
            buffers.append(i)
            stacks.append(stack[:])
            moves.append(best_move)
            stack,i,deps=self.transition(best_move,stack,i,deps)
            best_move=self.check_best(heads,stack,buffer,deps,i)
        if(i>len(buffer)):
            return stacks,buffers,moves
        else: return None




In [88]:
def encode_moves(oracle,parser,heads,phrase,phrases_lemma):
    stacks,buffers,moves=parser.simulate_parse(heads,phrase)
    embedded_features=oracle.extract_features(phrase,phrases_lemma,stacks,buffers)


    expanded_moves=[]
    for move in moves:
        if(move==0): expanded_moves.append(torch.tensor([1,0,0]))
        if(move==1): expanded_moves.append(torch.tensor([0,1,0]))
        if(move==2): expanded_moves.append(torch.tensor([0,0,1]))

    return embedded_features,expanded_moves

#prende il dataset e per ogni frase genera il batch corrispettivo, genererà n=batch-size file
def create_batches(oracle,parser,batch_size,dataset):
    batch_feature=[]
    batch_moves=[]

    index_batch=0
    for sent in dataset:
        if(index_batch<30):
            heads=[-1]
            words=[]
            lemmas=[]

            wrong_sent=0
            for token in sent:
                if(token.head is None): wrong_sent=1
                if(token.form is None): wrong_sent=1
                if(token.lemma is None): wrong_sent=1

                heads.append(int(token.head))
                words.append(token.form)
                lemmas.append(token.lemma)

            if(wrong_sent==0):
                sent_features,sent_moves = encode_moves(oracle,parser,heads,words,lemmas)
                sent_features=torch.stack(sent_features,dim=0).to(torch.float32)
                sent_moves=torch.stack(sent_moves).to(torch.float32)
                batch_feature.append(sent_features)
                batch_moves.append(sent_moves)


            if(len(batch_moves)==batch_size):
                #print(sent_moves)
                print(index_batch)
                packed_features=pack_sequence(batch_feature,enforce_sorted=False)
                packed_moves=pack_sequence(batch_moves,enforce_sorted=False)


                torch.save((packed_features,packed_moves), f"/content/drive/MyDrive/magistrale/TLN/batches/tensor{index_batch}.pt")
                index_batch+=1
                batch_feature=[]
                batch_moves=[]
#DA FARE:
#1. Salvare batch da 50-100 frasi su file dati, ossia per ogni frase gli stati con relativi stack,buffer e move.
#2. Importare batch per batch come nell'esempio di chat gpt.
#3. Per ogni batch fare forward e back-prop di adam optimizer come nell'esempio di chat_gpt.



#p1,m1=parser.encode_moves(heads,phrase,phrase)
#p2,m2=parser.encode_moves(heads2,phrase2,phrase2)
#print(m2)
#torch.set_printoptions(profile="full")
#p1 = torch.stack(p1, dim=0)
#p2 = torch.stack(p2, dim=0)
#print(p1)
#print(p2)
#input=pack_sequence([p1, p2])

#input_size = 3840  # Each element in the sequence is a vector of size 2
#hidden_size = 64
#num_layers = 1
#output_size = 3  # Example output size
#model = torch.nn.LSTM(input_size,hidden_size,num_layers,batch_first=True,bidirectional=False,proj_size=output_size)
#model.half()
#
#output = model(input)
#print(output)
#print("OUTPUT")
#print(unpack_sequence(output[0]))


oracle = Oracle()
oracle.train_on_batches()
#create_batches(oracle,parser,100,train_prepocesed)



Loadind 0
Loadind 1
Loadind 2
Loadind 3
Loadind 4
Loadind 5
Loadind 6
Loadind 7
Loadind 8
Loadind 9
Loadind 10
Loadind 11
Loadind 12
Loadind 13
Loadind 14
Loadind 15
Loadind 16
Loadind 17
Loadind 18
Loadind 19
Loadind 20
Loadind 21
Loadind 22
Loadind 23
Loadind 24
Loadind 25
Loadind 26
Loadind 27
Loadind 28
Loadind 29
Loadind 0
Loadind 1
Loadind 2
Loadind 3
Loadind 4
Loadind 5
Loadind 6
Loadind 7
Loadind 8
Loadind 9
Loadind 10
Loadind 11
Loadind 12
Loadind 13
Loadind 14
Loadind 15
Loadind 16
Loadind 17
Loadind 18
Loadind 19
Loadind 20
Loadind 21
Loadind 22
Loadind 23
Loadind 24
Loadind 25
Loadind 26
Loadind 27
Loadind 28
Loadind 29
Loadind 0
Loadind 1
Loadind 2
Loadind 3
Loadind 4
Loadind 5
Loadind 6
Loadind 7
Loadind 8
Loadind 9
Loadind 10
Loadind 11
Loadind 12
Loadind 13
Loadind 14
Loadind 15
Loadind 16
Loadind 17
Loadind 18
Loadind 19
Loadind 20
Loadind 21
Loadind 22
Loadind 23
Loadind 24
Loadind 25
Loadind 26
Loadind 27
Loadind 28
Loadind 29
Loadind 0
Loadind 1
Loadind 2
Loadind 3


In [89]:


parser = Parser(oracle)
phrases1= "Si chiamava Prototaxites .".split()
phrases2= "In la mia mente , almeno , è così .".split()
phrases3=" I turisti possono usare le barche per visitare alcune zone di Roma .".split()
phrases4="Il più largo organismo a il mondo è un tappeto miceliale , di lo spessore di una parete cellulare .".split()
heads1=[-1,2,0,2,2]
heads2=[-1,4,4,4,9,4,4,6,9,0,9]
heads3=[-1,2,4,4,0,6,4,8,4,10,8,12,10,4]
heads4=[-1,4,3,4,0,7,7,4,4,10,4,4,4,15,15,4,18,18,15,18,4]
deps = parser.parsing(phrases3,phrases3)
print(deps.get_heads())
stacks,buffers,moves=parser.simulate_parse(heads3,phrases3)
print(stacks)
print(moves)




[None, 2, 3, 4, 0, 6, 4, 8, 4, 10, 8, 12, 10, 4]
[[0], [0, 1], [0, 1, 2], [0, 2], [0, 2, 3], [0, 2, 3, 4], [0, 2, 4], [0, 4], [0, 4, 5], [0, 4, 5, 6], [0, 4, 6], [0, 4], [0, 4, 7], [0, 4, 7, 8], [0, 4, 8], [0, 4, 8, 9], [0, 4, 8, 9, 10], [0, 4, 8, 10], [0, 4, 8, 10, 11], [0, 4, 8, 10, 11, 12], [0, 4, 8, 10, 12], [0, 4, 8, 10], [0, 4, 8], [0, 4], [0, 4, 13], [0, 4]]
[0, 0, 2, 0, 0, 2, 2, 0, 0, 2, 1, 0, 0, 2, 0, 0, 2, 0, 0, 2, 1, 1, 1, 0, 1, 1]


In [90]:
def accuracy_on_test():
  #Load e preprocessing
  test = pyconll.load_from_file('/content/drive/MyDrive/magistrale/TLN/data/it_isdt-ud-test.conllu')
  test_prepocesed=[]
  for i,sent in enumerate(test):
      sentence_preprocesed=[]
      for j,token in enumerate(sent):
          if(token.head is not None):
              sentence_preprocesed.append(token)
      test_prepocesed.append(sentence_preprocesed)

  #Accuracy
  total_tokens=0
  total_wrong_tokens=0
  for i,sent in enumerate(test):
    if(i<100):
      print(i)
      heads=[-1]
      words=[]
      lemmas=[]

      wrong_sent=0
      for token in sent:
        if(token.head is None): wrong_sent=1
        else:
          heads.append(int(token.head))
        if(token.form is None): wrong_sent=1
        if(token.lemma is None): wrong_sent=1


        words.append(token.form)
        lemmas.append(token.lemma)

      if(wrong_sent!=1):
        predicted_deps=parser.parsing(words,lemmas)
        predicted_heads=predicted_deps.get_heads()
        sent_tokens=len(words)
        sent_wrong_tokens=0
        for j in range(1,len(predicted_heads)):
          if(predicted_heads[j] is not None):
            if(predicted_heads[j]!=heads[j]):
              sent_wrong_tokens+=1
          else:
            sent_wrong_tokens+=1
        total_tokens+=sent_tokens
        total_wrong_tokens+=sent_wrong_tokens

  accuracy=1-(total_wrong_tokens/total_tokens)
  return accuracy

print(accuracy_on_test())







0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
0.7773722627737226


hidden 64, epoch 1, lr 0.001, memory 0, bidir False, batches 30 0.7
hidden 64, epoch 1, lr 0.01, memory 0, bidir False, batches 30 0.55 overfitting
hidden 64, epoch 1, lr 0.0001, memory 0, bidir False, batches 0.58 underfitting

hidden 64, epoch 3, lr 0.001, memory 0, bidir False, batches 30 0.74
hidden 64, epoch 4, lr 0.001, memory 0, bidir False, batches 30 0.78 top
hidden 64, epoch 4, lr 0.001, memory 3, bidir False, batches 30 0.64 !

hidden 64, epoch 4, lr 0.001, memory 0, bidir False, batches 60 0.75  overfitting
hidden 64, epoch 3, lr 0.001, memory 0, bidir False, batches 60 0.76  overfitting
hidden 64, epoch 2, lr 0.001, memory 0, bidir False, batches 60 0.78  uguale a prima
hidden 64, epoch 1, lr 0.001, memory 0, bidir False, batches 60 0.71  underfitting

hidden 64, epoch 2, lr 0.0005, memory 0, bidir False, batches 120 0.78  
hidden 64, epoch 3, lr 0.0005, memory 0, bidir False, batches 120 0.78  
hidden 128, epoch 3, lr 0.0005, memory 0, bidir False, batches 120 0.815  top
hidden 512, epoch 4, lr 0.0005, memory 0, bidir False, batches 120 0.83  top